arch/arm/core/aarch64/macro_priv.inc

```diff
@@ -111,17 +112,56 @@
 	ldp	x0, x1, [sp], #16
 
 	/*
-	 * In general in the  ELR_ELn register we can find:
+	 * In general in the ELR_ELn register we can find:
 	 *
-	 * - The address of ret in z_arm64_call_svc() in case of arch_swap()
-	 *   (see swap.c)
 	 * - The address of the next instruction at the time of the IRQ when the
 	 *   thread was switched out.
 	 * - The address of z_thread_entry() for new threads (see thread.c).
 	 */
 	eret
 .endm
 
+/**
+ * @brief Save volatile registers for cooperative switching
+ *
+ * This is the equivalent of z_arm64_enter_exc() but used for cooperative switching
+ * in z_arm64_svc().  In this case we do not need to save all the x0-x18
```

Comment:
This seems right in the current context where SVC is being used to mimic a pendSV and achieving the context switching.
However in a generic SVC handler we need to save-restore all the registers like any other exception right, compiler will not do anything for the SVC instruction itself  ?

Reply:
The compiler will not do anything for the SVC instruction itself, it's up to us saving the registers. We need to preserve all the registers only for the preemptive context switch. For the cooperative switch it's not needed to preserve all the registers. The SVC call is only used for cooperative context switch or IRQ offloading, in both cases we only need a small subset of the registers saved, in all the other cases we are triggering a fatal error. In fact we currently do not have a generic SVC call handling code in place, we just trigger an error when we do not recognize a valid SVC parameter.

Please also note that this PR is a bit more than a cleanup. This commit is a preliminary work for the switch to the new `arch_switch()` API (that I have ready already) where the `x0` and `x1` registers content passed to `z_arm64_svc()` are being actively used in the thread switch code.

Reply:
Rest of the changes in the PR are clear to me. +1  Only this optimization of skipping save restore of corruptible registers in SVC handler is bit unconventional.  This skipping save restore is presumably not a necessary step for the next PR you mentioned.  I do agree there are no generic SVC usage as of now. But it can be in future. That time someone will have to add these back again.

And what happens to the irq offload case. The offload handlers may corrupt the x0-x18 ? Or they don't.  Just making sure.
I have not really looked into that path.

Reply:
I think the reason we don't need save x0-x18 is that there is no use x0-x18 within function z_arm64_call_svc, especially after `svc` instruction there is no use these caller-saved registers.
```
GTEXT(z_arm64_call_svc)
SECTION_FUNC(TEXT, z_arm64_call_svc)
        svc     #_SVC_CALL_CONTEXT_SWITCH
        ret
```

similar reason for z_arm64_offload.

Reply:
> And what happens to the irq offload case. The offload handlers may corrupt the x0-x18 ? Or they don't. Just making sure.

x0-x18 are caller-saved registers, that means that it is the caller's responsibility to push these registers onto the stack. This is true for the cooperative switch initiated with `arch_swap()` and for the IRQ offloading initiated with `arch_irq_offload()`. 

So, if I'm not wrong, each thread calling into `arch_swap()` or `arch_irq_offload()` already expects the registers x0-x18 to be corrupted eventually, so saving and restoring those is wasted work.

Reply:
>x0-x18 are caller-saved registers, that means that it is the caller's responsibility to push these registers onto the stack.

Compiler knows function like `z_arm64_call_svc` do not clobber all those registers but that is not true because  `svc `can.
I think weather it will affect something depends on how its used.  In this particular case of 'arch_awap()' (how its coded today) it looks safe and it works for sure.

Since other OS generally do save restore on a svc exception not doing so needs conformance in general. So its up to you to drop this svc handler save restore optimization part or keep it if confident about compiler behavior.

Reply:
> Compiler knows function like `z_arm64_call_svc` do not clobber all those registers but that is not true because `svc `can.

The SVC handler can clobber the caller-saved registers but that wouldn't matter to the calling function.

@sandeepbrcm anyway, I'll drop the second commit. I'll add a new cleanup commit so we can finally merge this (hopefully) and I can push the other stuff I have ready. Thanks for the review.